### ISSUE TITLE
Implementing a .checkovignore File to sets list of ruleID for Skip-Check with expiry Dates

### ISSUE THREAD

**Issue by iheanachochukwu** [bridgecrewio/checkov#6536]
## Summary
This feature request proposes the implementation of a .checkovignore file for Checkov to hold rule IDs that we desire to skip, along with their expiry dates and optional reasons. This functionality would be similar to the .trivyignore file used in Trivy, allowing users to specify the checks they want to ignore, set expiration dates for these checks, and optionally provide reasons for the skips.

## Background
Checkov is a powerful tool for detecting potential security and compliance misconfigurations in Infrastructure as Code (IaC). However, managing skip-checks directly can be cumbersome, especially for projects with a long list of checks to skip. A .checkovignore file would provide a centralized and manageable way to handle these skip-checks.

The .checkovignore file would allow users to specify rule IDs to ignore, set expiry dates for these skips, and optionally add reasons for skipping the checks. This can be particularly useful if a vulnerability does not affect your development environment or if there are other valid reasons to temporarily ignore certain checks.

## Proposal 

- **Create a .checkovignore File**:
The file will be in YAML format and include details about each check to be skipped, such as the check ID, expiry date, and an optional reason for skipping the check.

```yaml
# CHECKOV_IGNORE.py
import yaml
from datetime import datetime

def is_expired(expiry_date):
    current_date = datetime.now().date()
    expiry_date = datetime.strptime(expiry_date, "%Y-%m-%d").date()
    return current_date > expiry_date

try:
    # Load the YAML content from the file
    with open('.checkovignore', 'r') as file:
        data = yaml.safe_load(file)

    # Extract the 'id' values from the ignore list that are not expired
    ids = [rule['id'] for rule in data['ignore'] if not is_expired(rule['expiry'])]

    if ids:
        # Join the ids into a comma-separated string
        skip_checks = ','.join(ids)
    else:
        skip_checks = "null"

    # Get the total number of ids
    total_ids = len(ids)

    # Print the results in a way that can be captured by the shell script
    print(f"SKIP_CHECKS: {skip_checks}")
    print(f"TOTAL_SKIPPED: {total_ids}")

except FileNotFoundError:
    # Handle the case where the file is not found
    skip_checks = "null"
    print("[INFO], Checkov ignore file not found, assign SKIP_CHECKS as null, for error handling when running checkov scan...")
```

- **Python Script to Process .checkovignore**:
A Python script will read the .checkovignore file, filter out expired rules, and set the --skip-check parameter dynamically.

```yaml
# CHECKOV_IGNORE.py
import yaml
from datetime import datetime

def is_expired(expiry_date):
    current_date = datetime.now().date()
    expiry_date = datetime.strptime(expiry_date, "%Y-%m-%d").date()
    return current_date > expiry_date

try:
    # Load the YAML content from the file
    with open('.checkovignore', 'r') as file:
        data = yaml.safe_load(file)

    # Extract the 'id' values from the ignore list that are not expired
    ids = [rule['id'] for rule in data['ignore'] if not is_expired(rule['expiry'])]

    if ids:
        # Join the ids into a comma-separated string
        skip_checks = ','.join(ids)
    else:
        skip_checks = "null"

    # Get the total number of ids
    total_ids = len(ids)

    # Print the results in a way that can be captured by the shell script
    print(f"SKIP_CHECKS: {skip_checks}")
    print(f"TOTAL_SKIPPED: {total_ids}")

except FileNotFoundError:
    # Handle the case where the file is not found
    skip_checks = "null"
    print("[INFO], Checkov ignore file not found, assign SKIP_CHECKS as null, for error handling when running checkov scan...")
```

- **Example of Implementation in Azure DevOps Pipeline:**
Here's how I used the .checkovignore file in an Azure DevOps pipeline to dynamically set the skip-checks.
```yaml
steps:
  - script: |
      python CHECKOV_IGNORE.py
    displayName: 'Process .checkovignore file'

  - script: |
      checkov --directory . --skip-check $(SKIP_CHECKS)
    displayName: 'Run Checkov scan'
```

## Benefits
- Centralized Management: The .checkovignore file allows for centralized management of skip-checks, making it easier to handle multiple projects.

- Automated Expiry Handling: The Python script ensures that expired skip-checks are automatically excluded, promoting regular review and maintenance of security policies.

- Optional Reason Documentation: Provides context for why certain checks are skipped, aiding in auditing and future reviews.

## Conclusion

Implementing a .checkovignore file with expiry dates would significantly enhance the usability and manageability of Checkov. This feature will streamline the process of handling skip-checks, making it more efficient and maintainable for development teams.

I am willing to contribute this feature and look forward to any feedback or suggestions from the community on how to get started.

**Comment by iheanachochukwu** [comment]
@ChanochShayner @Saarett 
I've added this issue to propose and contribute the .checkovignore file feature for managing skip-checks with expiry dates. Looking forward to your feedback on how to proceed.

**Comment by Saarett** [comment]
Thank you @iheanachochukwu for the proposal!
@tsmithv11 please take a look at this proposal

**Comment by tsmithv11** [comment]
@iheanachochukwu thanks for this! Are you open to contributing it? If so, please provide tests to confirm it won't conflict with or cause problems when different flags are used.

**Comment by iheanacho-chukwu** [comment]
@tsmithv11 Thanks for replying. I've implemented the .checkovignore functionality in my Azure DevOps pipeline by using the above Python script that processes check IDs from the .checkovignore file and passes them to the Checkov command via --skip-check.

I can provide test that shows this works, however, I believe this feature could be more user-friendly if it were natively integrated into Checkov, removing the need for users to manage the Python script themselves. I’m open to contributing this directly into the Checkov codebase but would appreciate some guidance on how best to do so, particularly since this is my first contribution to the project.

Is there collaboration platform like slack where I can discuss this, it would be great to connect there for more interactive support.

**Comment by iheanacho-chukwu** [comment]
@tsmithv11 @Saarett 

A working example of Azure DevOps pipepline file- 
https://github.com/iheanacho-chukwu/checkov/blob/issue_6536_example_checkovignore_skips_azuredevops/tests/azure_pipelines/examples/azure-pipelines-checkov-ignore.yml

Azure Pipeline run result - https://dev.azure.com/nacho-chukwu/Infrastructure%20As%20A%20Code/_build/results?buildId=35&view=logs&j=734f1c92-bc4e-59a3-6d82-326d022426b0&t=3a2f8c00-e87e-51f7-e1ee-2eea722d0b64 
 ignoring the checkov scans for CKV_AZURE_35 and CKV_AZURE_43 using the [.checkovignore](https://github.com/iheanacho-chukwu/checkov/blob/issue_6536_example_checkovignore_skips_azuredevops/checkovignore/.checkovignore) file

**Comment by vish-space** [comment]
Hi @iheanacho-chukwu , just checking if this has been implemented yet ? - looking forward to this feature. Thankyou

**Comment by tsmithv11** [comment]
@iheanacho-chukwu I don't believe you need an invite, just sign up https://slack.bridgecrew.io/. Please note, that this feature is not prioritized right now as we have other pressing issues.
